### ISSUE TITLE
Device plugins work tracker

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#53497]
This issue is meant to track planned work items for device plugins across releases.
Device plugins enable compute resource extensibility at the node level.

This is **not finalized yet**. This issue will be updated once the community agrees to this schedule

## v1.9
[ ] - Avoid passing env vars, mounts and devices per device per Allocation. PR #53031 for background [**Owner**: @vishh]
[ ] - Stub E2E device plugin to build conformance testing for device plugin infra in OSS. [**Owner** TBD]
[ ] - Test device plugin integration with stub e2e in scale and soak test suites (Simulates production scenarios) [**Owner** TBD]
[ ] - Finalize an internal software architecture for Device plugin support in Kubelet to improve reliability, testability and readability [**Owner**: @jiayingz ]
[ ] - Remove feature gate for Device plugins - Device plugins will not reach beta at-least until we have 4 implementations that inform the design of the API. Until then there are no guarantees against API compatibility [**Owner** TBD]
[ ] - Develop a device plugin user guide that provide production rollout guidelines as well [**Owner** TBD]
[ ] - Start code refactoring [**Owner** TBD]
[ ] - Post a short overview for supporting Solarflare NICs with device plugins [**Owner**: @vikaschoudhary16]
[ ] - Add Pod Spec to Allocate RPC [**Owner**: @vikaschoudhary16 ]

## v1.10
[ ] - Complete code refactoring [**Owner** TBD]
[ ] - Explore support for Infiniband [**Owner** TBD]
[ ] - Propose design for topology support in device plugins [**Owner** TBD]
[ ] - Propose design for supporting extended resources via Quota, limit range, and UI [**Owner** @vishh ] 

cc @jiayingz @RenaudWasTaken @jeremyeder @vikaschoudhary16 @ConnorDoyle @derekwaynecarr @mindprince @cmluciano

**Comment by lichuqiang** [comment]
If no one is doing it, I would like to work on the code refactoring part.

**Comment by vishh** [comment]
@lichuqiang Thanks for volunteering. That will have to wait on @jiayingz's design doc for the internal architecture though. Without an approved plan, refactoring may not end up yielding the value necessary. 

Meanwhile, are you interested in picking up any other work items?

**Comment by ConnorDoyle** [comment]
> Propose design for topology support in device plugins

I'd like to help gather everyone's requirements on this and work toward a design.

**Comment by lichuqiang** [comment]
> Allocate devices during pod creation rather than during every container (re)start

@vishh This makes sense. I'd like to help with it.
Besides. I have been working on the checkpoint issue #52678
I think  implementation of the issue depends on modification in Allocate behavior as well :)

**Comment by sameo** [comment]
> Allocate devices during pod creation rather than during every container (re)start [Owner TBD]
@vishh Makes sense to me. I'd like to work on that one if that's fine with you.

**Comment by sameo** [comment]
@vishh For v1.10 I would like to propose adding:

Explore support for FPGA [Owner @sameo]

**Comment by dchen1107** [comment]
Can someone explain why we want to pass Pod Spec to Allocate RPC? I have concerns on passing the entire pod spec to plug-in in general due to the maintenance unknown. The same issue was raised when we first design CRI. cc/ @kubernetes/sig-node-feature-requests

**Comment by RenaudWasTaken** [comment]
> Start code refactoring

Related:
- Issue #51993
- PR #52747
- PR #52616
- PR #52149
- PR #52147

**Comment by resouer** [comment]
@ConnorDoyle I really have some inputs on topology design from research side (and I've once talked it with @renaudwastaken), would like to participate together on this task.

**Comment by ScorpioCPH** [comment]
@vishh I think i can pick up `stub e2e` work items, and i am trying POC implementation of [container-engine-accelerators](https://github.com/GoogleCloudPlatform/container-engine-accelerators) on ubuntu right now :-)

**Comment by resouer** [comment]
> Allocate devices during pod creation rather than during every container (re)start [Owner TBD]

@sameo Will this work involve CRI change?

**Comment by sameo** [comment]
> Will this work include CRI change?

@resouer I think so, yes. I believe it would make sense to do the device plugin allocations and the runtime ones at the same time, i.e. at Pod creation time. So we should have a `Devices` field in `PodSandboxConfig` and maybe remove the one in `ContainerConfig`.

**Comment by lichuqiang** [comment]
@vishh Okey, I‘ll keep tracking the refactoring job.
For now, I'd like to continue working on some issues left in v1.8. Say _Checkpoint artifacts issue_ #52678, _admission_ issue #51592
And take part in some test work if necessary :)

**Comment by resouer** [comment]
@dchen1107 Hi Dawn, just discussed with folks in wg today. 

1. For Solarflare NIC support, there's proposal that Pod metadata like name is needed to find correct net ns of pod.
2. For device topology support, pod topology requirement is expected to be passed to Allocate(), but we may also be able to use annotation instead in this case.

I think @vishh may have more inputs here.

**Comment by ConnorDoyle** [comment]
@resouer want to take up the topology discussion in #49964 comments?

**Comment by jiayingz** [comment]
Can we create separate issues (if an issue hasn't been created yet) to discuss individual topics instead of overloading this issue with mixed discussions on different topics? It is a bit hard to follow at this point.

**Comment by resouer** [comment]
@ConnorDoyle Yes, I will DM you with some draft.

**Comment by vishh** [comment]
+1 for creating separate issues for each of the topics here. This issue is meant to be a tracker for lack of a better tool in github.

**Comment by vishh** [comment]
@ScorpioCPH I have added you as an owner for creating stub device plugin for e2es. Please open an issue with your design for a stub device plugin before you post a patch to help with the review process. 
Keep in mind, there's only ~4 more weeks left in v1.9 for development

**Comment by vishh** [comment]
@resouer on https://github.com/kubernetes/kubernetes/issues/53497#issuecomment-336093773

Based on offline ping, if/when kubelet supports updating a pod with a new container, I'd expect feasibility checks and admission logic to be re-run. All of the container manager logic has to be re-run at which point I'd expect additional devices to be allocated for new container(s) being added to the pod. 
What I'm expecting is only one idempotent API from the container manager that the kubelet will invoke which will be `EnsureExists()` or something similar which will create all dependencies like pod cgroups, cpuset allocations, devices (including extended), etc. 

cc @jiayingz

**Comment by vishh** [comment]
@sameo Looks like @lichuqiang is already working on resolving # #51592
https://github.com/kubernetes/kubernetes/pull/51895

Will you be able to help with scale and soak testing meanwhile? You'll get a good handle on k8s test infra and device plugins in that process and help stabilize the existing implementation.

**Comment by tengqm** [comment]
Related Issue: #51999

**Comment by ScorpioCPH** [comment]
@vishh @jiayingz [[Proposal] Stub Device Plugin for E2E Test](https://github.com/kubernetes/kubernetes/issues/54412), PTAL.

**Comment by pineking** [comment]
@vishh @cmluciano could we add this feature request https://github.com/kubernetes/kubernetes/issues/52412 into milestone v1.10 or v1.11?

**Comment by ScorpioCPH** [comment]
PR: [Add stub device plugin for conformance e2e test](https://github.com/kubernetes/kubernetes/pull/54792)

**Comment by CodeJuan** [comment]
so

**Comment by choury** [comment]
I think  we should consider the interconnect performance between the GPUs and CPUs to minimize latency.
The  distance map of GPUs can be got through nvml library  since google/cadvisor#1436 has merged.

**Comment by resouer** [comment]
@choury Cloud you please explain a little about `distance map of GPUs` you are referring to?

**Comment by jiayingz** [comment]
@choury I think https://github.com/google/cadvisor/pull/1762 only exposes basic gpu metrics not including gpu-to-cpu latency metrics. If you think it is important to take gpu-to-cpu latency metric into account, could you write a document describing the important use cases you would like to support? Supporting this kind of advanced scheduling is perhaps beyond the device plugin scope but may fit into the resource api work we plan to tackle next year.

**Comment by RenaudWasTaken** [comment]
@jiayingz see [this old design doc](https://docs.google.com/document/d/1lSwVh2ZfJ2FeLXIeyyiNqN_hKPYpahJiwN5X5cszjOk/edit#heading=h.wj7s0tu10uz2) that explains problems and use cases.

Basically if gpu-to-cpu latency / topology is not taken into account you might as well take a CPU only instance.

> Supporting this kind of advanced scheduling is perhaps beyond the device plugin scope but may fit into the resource api work we plan to tackle next year.

Since you're mentioning it I'm all for supporting this as the resource level.

**Comment by flx42** [comment]
Topology and affinity can be queried like this:
```
$ nvidia-smi topo -m                                                                                                                                                                 
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    mlx4_0  CPU Affinity                                                                                                 
GPU0     X      PIX     PHB     PHB     SYS     SYS     SYS     SYS     PHB     0-9,20-29                                                                                                    
GPU1    PIX      X      PHB     PHB     SYS     SYS     SYS     SYS     PHB     0-9,20-29                                                                                                    
GPU2    PHB     PHB      X      PIX     SYS     SYS     SYS     SYS     PHB     0-9,20-29                                                                                                    
GPU3    PHB     PHB     PIX      X      SYS     SYS     SYS     SYS     PHB     0-9,20-29                                                                                                    
GPU4    SYS     SYS     SYS     SYS      X      PIX     PHB     PHB     SYS     10-19,30-39                                                                                                  
GPU5    SYS     SYS     SYS     SYS     PIX      X      PHB     PHB     SYS     10-19,30-39                                                                                                  
GPU6    SYS     SYS     SYS     SYS     PHB     PHB      X      PIX     SYS     10-19,30-39                                                                                                  
GPU7    SYS     SYS     SYS     SYS     PHB     PHB     PIX      X      SYS     10-19,30-39                                                                                                  
mlx4_0  PHB     PHB     PHB     PHB     SYS     SYS     SYS     SYS      X                                                                                                                   
                                                                                                                                                                                             
Legend:                                                                                                                                                                                      
                                                                                                                                                                                             
  X    = Self                                                                                                                                                                                
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)                                                                                       
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe switches (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing a single PCIe switch
  NV#  = Connection traversing a bonded set of # NVLinks
```
Scheduling GPU 0 with cores 10-19 would not be good for performance.
Scheduling GPU 0 and GPU 7 for a multi-GPU job would not be good for performance.

**Comment by jeremyeder** [comment]
We need to handle NUMA and PCI locality in a non-GPU specific way.  I am not sure this issue is the right place to drag in this additional requirement.  @ConnorDoyle would you point to your PCI/NUMA tracker?

**Comment by ConnorDoyle** [comment]
https://github.com/kubernetes/kubernetes/issues/49964 Hardware topology awareness at node level (including NUMA)

**Comment by choury** [comment]
@resouer It should be called distance matrix? like @flx42  shows.
@jiayingz Yes,  but google/cadvisor#1762 introduced the nvml library, We can get this information use [nvmlDeviceGetTopologyNearestGpus](https://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html#group__nvmlDeviceQueries_1g6c812383e9505c34eee61f8a28939f6f) or [nvmlDeviceGetTopologyCommonAncestor](https://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html#group__nvmlDeviceQueries_1g518f3f171c40e5832cb2b1b26e960b2b)  .  Not only the gpu-to-gpu   but the gpu-to-cpu  latency should be considered, so k8s should provides CPU or NUMA information to  device plugins.

**Comment by resouer** [comment]
@choury @flx42 @jeremyeder Thanks! So it sounds like a use case for #49964 , we can move future discussion to there if any. 

While I agree with @jiayingz this is indeed more than current device plugin to handle so please expect long term work and cooperation here. 

As we are currently at requirement gathering phase. @ConnorDoyle (owner of the issue) and me have worked out a [public doc](https://docs.google.com/document/d/1dv3-se21ivEZhyRCr9qixvjyB_VRO2yIlotWb2kOdiw/edit?ts=59e4be43#) to collect these topology related requirements, so please feel free to describe yours and add it into the doc. It would be a good start!

**Comment by jiayingz** [comment]
Here is a brief update on the 1.9 Device Plugin work:
- Device Plugin API changes to v1alpha2 with AllocateResponse changed to list aggregated artifacts (PR 53031) Done
- Certain local cleanups (PR 52147 PR 52747) Done
- Device plugin code now locates in a single directory under pkg/kubelet/cm/deviceplugin (PR 54019). Done
- Better handling of device plugin resource deletion (Issue 53395 PR 54488 PR 55088). Done.
- Merge Handler functionalities into Manager and clean up deviceplugin public interface (issue 55180 PR 55730). Done
- Uses file store utility for device plugin checkpointing (PR 55382). Hopefully to get merged soon.
- Remove device plugin feature gate (PR 55828) TBD

There is also work-in-progress on areas like stub device plugin based e2e tests, continue code refactoring etc, which will probably be carried over to 1.10.

**Comment by andyxning** [comment]
Too long thread. But, what is the link for the design docs @jiayingz

**Comment by jiayingz** [comment]
@andyxning https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md

**Comment by bhack** [comment]
The majority of subtasks have 1.9 as target. Can you refactor this tracking issue for 1.10/1.11?

**Comment by BSWANG** [comment]
Device plugin on network interface schedule. #56634 
Current resources for device plugins allocated are container level. But the network interfaces are pod/sandbox level. Is the device plugin suiting on network device?

**Comment by fabiand** [comment]
@BSWANG true - After a few months I really wonder if this is a _real_ issue atm.

The bigger problem is probably the life-cycle events which are currently container focused.
